### PR TITLE
[PDI-18375] Implement Prefetch message count for streaming steps

### PR DESCRIPTION
**Pull request by ricksjames**

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 43m
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
engine,plugins/streaming/impls/jms,plugins/streaming/impls/mqtt,ui
```
---
:ok_hand: All tests passed!
 
Tests run: **3660**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7345/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 2m 49s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
engine,plugins/streaming/impls/jms,plugins/streaming/impls/mqtt,ui
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :x: Build failed in 2m 53s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
engine,plugins/streaming/impls/jms,plugins/streaming/impls/mqtt,ui
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 41m 26s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
engine,plugins/streaming/impls/jms,plugins/streaming/impls/mqtt,ui
```
---
:ok_hand: All tests passed!
 
Tests run: **3660**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7345/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Atomic Count Semaphore
+ * <br>
+ * This class implements just a subset of the functionality of {@link Semaphore} but provides faster performance. Also,
+ * {@link Semaphore#release(int)} in testing was found not to always be thread safe.
+ *
+ * Note that this implementation has a very narrow usage scope with the following restrictions:
+ * 1. Acquire may only be called by one thread, otherwise more permits can be acquired than permitCount
```
Comment (mkambol) [inline engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java]: Is it the thread that *creates* the semaphore that is allowed to call `.acquire`?  If so, should lock that down with a precondition.

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+   * @param count
+   * @return the permit count after the release
+   */
+  public int release( int count ) {
+    //release the semaphore if the count is more than 0 and the permit count was zero before the add
+    int afterAddCount = 0;
+    if ( count > 0 ) {
+      if ( ( afterAddCount = permitCount.getAndAdd( count ) ) == 0 ) {
+        this.semaphore.release();
+      }
```
Comment (mkambol) [inline engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java]: The above conditional is not an atomic action, so if 2 or more threads are releasing there's the possibility of semaphore.release being called extra times.  Scenario:
1) T1 `acquire()`, permitCount set to 0
2) T2 `release(1)`, sees permit count of 0, increments to 1.  Gets past line 56, but not yet to line 57.
3) T1 `acquire()`, decrements permit count to 0, `Semaphore.acquire`, blocks waiting for semaphore 
4) T3 `release(1)`, sees a permit count of 0, increments to 1, gets past line 56.
5) T2 and T3 both release the semaphore, increasing its Semaphore.permit count to 2.

This also gets a bit muddled to talk about with this class using the acquire/permit/release terminology, which is subtly different then the acquire/permit/release terminology used by the underlying semaphore.

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/BaseStreamStep.java [hunk 3]
```diff
@@ -189,6 +189,19 @@ protected int getBatchSize() {
...
+  /**
+   * Get Prefetch Count
+   *
+   * @return the number of messages to prefetch from the broker
+   */
+  protected int getPrefetchCount() {
+    try {
+      return Integer.parseInt( variablizedStepMeta.getPrefetchCount() );
+    } catch ( NumberFormatException nfe ) {
+      return 100000;
```
[the remark this answers is not in the document]
Comment (ricksjames) [reply engine/src/main/java/org/pentaho/di/trans/streaming/common/BaseStreamStep.java]: fixed

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/BaseStreamStepMeta.java [hunk 4]
```diff
@@ -166,6 +182,27 @@ public void check( List<CheckResultInterface> remarks, TransMeta transMeta,
...
+    try {
+      int prefetch = Integer.parseInt( space.environmentSubstitute( getPrefetchCount() ) );
+
+      if ( prefetch <= 0 ) {
+        remarks.add( new CheckResult( CheckResultInterface.TYPE_RESULT_ERROR,
+          BaseMessages.getString( PKG, "BaseStreamStepMeta.CheckResult.PrefetchGreaterThanZero", prefetch, size ),
+          stepMeta ) );
+      }
+
+      if ( prefetch <= size ) {
```
[the remark this answers is not in the document]
Comment (ricksjames) [reply engine/src/main/java/org/pentaho/di/trans/streaming/common/BaseStreamStepMeta.java]: updated wording.

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/BlockingQueueStreamSource.java [hunk 5]
```diff
@@ -100,29 +99,42 @@ protected BlockingQueueStreamSource( BaseStreamStep streamStep ) {
...
+        streamStep.getSubtransExecutor().acquireBufferPermit();
         streamStep.incrementLinesInput();
         publishProcessor.onNext( row );
-      } );
+      }
     } catch ( InterruptedException e ) {
-      logChannel.logError(
-        getString( PKG, "BlockingQueueStream.AcceptRowsInterrupt",
-          Arrays.toString( rows.toArray() ) ) );
+      logChannel
```
[the remark this answers is not in the document]
Comment (ricksjames) [reply engine/src/main/java/org/pentaho/di/trans/streaming/common/BlockingQueueStreamSource.java]: fixed

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/BlockingQueueStreamSource.java [hunk 6]
```diff
@@ -100,29 +99,42 @@ protected BlockingQueueStreamSource( BaseStreamStep streamStep ) {
...
+  /**
+   * Wait for Subscribers Wait for the publish processor to have subscribers, otherwise the subtrans misses messages
+   * (even in .onSubscribe() the .hasSubscribers() is still false for a short while)
+   *
+   * @throws InterruptedException
+   */
+  private void waitForSubscribers() throws InterruptedException {
+    while ( !publishProcessor.hasSubscribers() ) {
+      Thread.sleep( 100 );
+      logChannel.logDebug( getString( PKG, "BlockingQueueStream.WaitForSubscribers" ) );
```
[the remark this answers is not in the document]
Comment (ricksjames) [reply engine/src/main/java/org/pentaho/di/trans/streaming/common/BlockingQueueStreamSource.java]: Moved the log to before and after the while loop so it won't flood, but could still be helpful.

File: engine/src/test/java/org/pentaho/di/trans/SubtransExecutorTest.java [hunk 7]
```diff
@@ -197,4 +205,79 @@ public void doesNotExecuteWhenStopped() throws KettleException {
...
+
+  /**
+   * @param future
+   * @return false if timed out, otherwise true
+   */
+  private boolean safeWaitForCompletion( Future future, long timeoutSeconds ) {
+    try {
+      future.get( timeoutSeconds, TimeUnit.SECONDS );
+    } catch ( TimeoutException te ) {
+      System.out.println( "Thread timed out." );
```
[the remark this answers is not in the document]
Comment (ricksjames) [reply engine/src/test/java/org/pentaho/di/trans/SubtransExecutorTest.java]: removed

File: engine/src/test/java/org/pentaho/di/trans/SubtransExecutorTest.java [hunk 8]
```diff
@@ -197,4 +205,79 @@ public void doesNotExecuteWhenStopped() throws KettleException {
...
+    List<RowMetaAndData> rows = Arrays.asList(
+      new RowMetaAndData( rowMeta, "Pentaho", 1L ),
+      new RowMetaAndData( rowMeta, "Pentaho", 2L ),
+      new RowMetaAndData( rowMeta, "Pentaho", 3L ),
+      new RowMetaAndData( rowMeta, "Pentaho", 4L ) );
+
+    Future future = executorService.submit( () -> safeAcquirePermits( subtransExecutor, 8 ) );
+
+    //Acquire the permits before releasing them by waiting briefly for the thread to spin up
+    Thread.sleep( 1000 );
```
[the remark this answers is not in the document]
Comment (ricksjames) [reply engine/src/test/java/org/pentaho/di/trans/SubtransExecutorTest.java]: implemented with CompletableFuture

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  /**
+   * Release
+   *
+   * @param count
+   * @return the permit count after the release
+   */
+  public int release( int count ) {
+    //release the semaphore if the count is more than 0 and the permit count was zero before the add
```
[the remark this answers is not in the document]
Comment (ricksjames) [reply engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java]: removed class

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java (hunk 1 above)
Comment (ricksjames) [reply engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java]: removed class

File: engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java (hunk 2 above)
Comment (ricksjames) [reply engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java]: removed class

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 42m 44s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
engine,plugins/streaming/impls/jms,plugins/streaming/impls/mqtt,ui
```
---
:ok_hand: All tests passed!
 
Tests run: **3653**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7345/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 41m 22s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
engine,plugins/streaming/impls/jms,plugins/streaming/impls/mqtt,ui
```
---
:ok_hand: All tests passed!
 
Tests run: **3653**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7345/6/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Pull request merged**